File: packages/dd-trace/test/llmobs/parity/scenarios/vertexai/_runner.js
'use strict'

const path = require('node:path')
const { finish, loadTracer } = require('../common')

const ROOT = path.resolve(__dirname, '../../../../../../../')
const sdkPackage = path.join(ROOT, 'versions/@google-cloud/vertexai')

async function run (scenario) {
  const tracer = loadTracer('google-cloud-vertexai')
  const vertex = require(sdkPackage).get()
  const { GoogleAuth } = require(path.join(
    ROOT,
    'versions/@google-cloud/vertexai/node_modules/google-auth-library/build/src/auth/googleauth'
  ))
  const originalToken = GoogleAuth.prototype.getAccessToken
  const originalFetch = global.fetch
  const provider = new URL(process.env.PROVIDER_BASE_URL)
  GoogleAuth.prototype.getAccessToken = async () => ({ token: 'test' })
  global.fetch = (url, options) => {
    const target = new URL(url)
    target.protocol = provider.protocol
    target.host = provider.host
    return originalFetch(target, options)
  }
  try {
    const client = new vertex.VertexAI({ project: 'parity-project', location: 'us-central1', apiEndpoint: provider.host })
    const model = client.getGenerativeModel({
      model: 'gemini-1.5-flash',
      systemInstruction: 'You are a parity bot.',
      generationConfig: { temperature: 1, maxOutputTokens: 50 },
      ...(scenario === 'generate-content-tools'
        ? {
            tools: [{
              functionDeclarations: [{
                name: 'add',
                description: 'Add two numbers.',
                parameters: { type: 'object', properties: { a: { type: 'number' }, b: { type: 'number' } } },
              }],
            }],
          }
        : {}),
    })
    if (scenario === 'generate-content-stream') {
      const result = await model.generateContentStream('Hello, how are you?')
      for await (const chunk of result.stream) void chunk
      await result.response
    } else if (scenario === 'chat-send-message') {
      const chat = model.startChat({
        history: [
          { role: 'user', parts: [{ text: 'Hello.' }] },
          { role: 'model', parts: [{ text: 'Hi.' }] },
        ],
      })
      await chat.sendMessage('Continue the parity conversation.')
    } else {
      await model.generateContent(scenario === 'generate-content-tools'
        ? {
            contents: [
              { role: 'user', parts: [{ text: 'What is 2 + 2?' }] },
              { role: 'model', parts: [{ functionCall: { name: 'add', args: { a: 2, b: 2 } } }] },
              { role: 'user', parts: [{ functionResponse: { name: 'add', response: { result: 4 } } }] },
            ],
          }
        : 'Hello, how are you?')
    }
  } catch (error) {
    void error
  } finally {
    GoogleAuth.prototype.getAccessToken = originalToken
    global.fetch = originalFetch
    await finish(tracer)
  }
}

module.exports = { run }

File: packages/dd-trace/test/llmobs/parity/scenarios/vertexai/_runner.py
import os
from urllib.parse import urlparse

import requests
import vertexai
from google.auth.credentials import AnonymousCredentials

from common import finish, start


def run(scenario):
    start("vertexai")
    provider = os.environ["PROVIDER_BASE_URL"]
    host = urlparse(provider).netloc
    original_request = requests.sessions.Session.request

    def request(self, method, url, **kwargs):
        return original_request(self, method, url.replace(f"https://{host}", provider), **kwargs)

    requests.sessions.Session.request = request
    try:
        vertexai.init(
            project="parity-project",
            location="us-central1",
            api_endpoint=host,
            api_transport="rest",
            credentials=AnonymousCredentials(),
        )
        from vertexai.generative_models import GenerativeModel
        from vertexai.generative_models import Content, Part
        from vertexai.generative_models import _generative_models as generative_models

        model = GenerativeModel(
            "gemini-1.5-flash",
            system_instruction="You are a parity bot.",
            generation_config={"temperature": 1, "max_output_tokens": 50},
        )
        if scenario == "generate-content-stream":
            response = model.generate_content("Hello, how are you?", stream=True)
            for _ in response:
                pass
        elif scenario == "chat-send-message":
            chat = model.start_chat(history=[
                Content(role="user", parts=[Part.from_text("Hello.")]),
                Content(role="model", parts=[Part.from_text("Hi.")]),
            ])
            chat.send_message("Continue the parity conversation.")
        elif scenario == "generate-content-tools":
            contents = [
                Content(role="user", parts=[Part.from_text("What is 2 + 2?")]),
                Content(role="model", parts=[generative_models.Part._from_gapic(generative_models.gapic_content_types.Part(function_call=generative_models.gapic_tool_types.FunctionCall(name="add", args={"a": 2, "b": 2})))]),
                Content(
                    role="user",
                    parts=[Part.from_function_response(name="add", response={"result": 4})],
                ),
            ]
            model.generate_content(contents)
        else:
            model.generate_content("Hello, how are you?")
    except Exception:
        pass
    finally:
        requests.sessions.Session.request = original_request
        finish()
